tiledb/bioimg/converters/base.py

```diff
@@ -224,13 +204,28 @@ def write_level_image(
         :param metadata: Metadata for the given level
         :param image_mask: Mask the original image depending on export format requirements
         """
+        ...
+
+
+# Define covariant type variables
+TReader = TypeVar("TReader", bound=ImageReader)
+TWriter = TypeVar("TWriter", bound=ImageWriter)
+
 
+class ImageConverter(Protocol[TReader, TWriter]):
+    _ImageReaderType: Optional[Type[TReader]] = None
+    _ImageWriterType: Optional[Type[TWriter]] = None
 
-class ImageConverter:
+    def from_tiledb(self) -> ImageConverter[TReader, TWriter]: ...
+
+    def to_tiledb(self) -> ImageConverter[TReader, TWriter]: ...
```

Comment:
I don’t quite follow what’s going on here. An ImageConverter has a method that returns another thing of its own type? This annotation would indicate to me that I could do this:

```python
converter = build_some_converter()
other_converter = converter.from_tiledb()
third_converter = converter.to_tiledb()
# and even
more_converter = third_converter.from_tiledb()
```

the type of all of which would be some form of `ImageConverter[...]`.

Based on the below, it looks like you might be intending to make these `@classmethod`s with at least a subset of the parameters below.